core/src/main/java/org/opensearch/sql/expression/function/VarargsFunctionResolver.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.expression.function;
+
+import java.util.AbstractMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Singular;
+import org.apache.commons.lang3.tuple.Pair;
+import org.opensearch.sql.exception.ExpressionEvaluationException;
+
+/**
+ * The Function Resolver hold the overload {@link FunctionBuilder} implementation.
+ * is composed by {@link FunctionName} which identified the function name
+ * and a map of {@link FunctionSignature} and {@link FunctionBuilder}
+ * to represent the overloaded implementation
+ */
+@Builder
+@RequiredArgsConstructor
+public class VarargsFunctionResolver implements FunctionResolver {
```

Comment:
This class is almost the same as `DefaultFunctionResolver`. Any chance we can delegate most logic to default resolver? As I understand, given source signature `concat(string, string, string) and vararg type is string, we only need to let default resolver work with candidate target signature `concat(string, string, string)` right? In this way, I think your resolver can work with existing type widen rule and implicit cast in function repository.

Reply:
Yes, in `DefaultFunctionResolver` I could add a check as shown in below code snippet, and it would work for `concat()`, but later, if we have other varargs functions (Ex: `concat_ws()` could be refactored too), then more checks will be needed to be added to `DefaultFunctionResolver`. 
```if (FunctionSignature.NOT_MATCH.equals(bestMatchEntry.getKey())
    && !unresolvedSignature.getFunctionName().equals(BuiltinFunctionName.CONCAT.getName())) {
      throw new ExpressionEvaluationException(
          String.format("%s function expected %s, but get %s", functionName,
              formatFunctions(functionBundle.keySet()),
              unresolvedSignature.formatTypes()
          ));
    }```
  Would you advice to delegate this to `DefaultFunctionResolver` as above or we can keep `VarargsFunctionResolver` as later we might have other varargs functions (like `concat_ws()` being refactored) as well?

Reply:
Have a look on team review and discussion https://github.com/Bit-Quill/opensearch-project-sql/pull/200#discussion_r1062921265 it could be helpful

Reply:
Will check and see if I missed anything. Thanks!

Reply:
As a reference. https://github.com/crate/crate/blob/3.3/sql/src/main/java/io/crate/expression/scalar/ConcatFunction.java#L127